src/tests/Microsoft.PowerFx.Connectors.Tests.Shared/SafeDTypeToStringTests.cs

```diff
@@ -0,0 +1,517 @@
+            Assert.Equal("*[field1:n, field2:b, field3:n]", type3.SafeDTypeToString());
+
+            type3 = type2.SetType(ref err, DPath.Root.Append(new DName("B")).Append(new DName("X")), DType.Boolean);
+            Assert.Equal("*[field1:n, field2:n, field3:n]", type3.SafeDTypeToString());
+
+            var type4 = type1.Add(ref err, DPath.Root.Append(new DName("X")), new DName("D"), DType.Number);
+
+            Assert.Equal("*[field1:n, field2:n, field3:n]", type4.SafeDTypeToString());
+
+            var type5 = DType.EmptyTable.Add(ref err, DPath.Root, new DName("D"), DType.String)
+                .Add(ref err, DPath.Root, new DName("E"), DType.String);
+
+            Assert.Equal("*[field1:s, field2:s]", type5.SafeDTypeToString());
+
+            var type6 = DType.CreateTable(new TypedName(type5, new DName("Y")), new TypedName(DType.Boolean, new DName("Q")));
+
+            Assert.Equal("*[field1:b, field2:*[field1:s, field2:s]]", type6.SafeDTypeToString());
+
+            var type = type6.Add(ref err, DPath.Root.Append(new DName("Y")), new DName("E"), DType.Error);
+
+            Assert.Equal("*[field1:b, field2:*[field1:s, field2:e]]", type.SafeDTypeToString());
+            err = false;
+
+            DType type7 = type6
+                .Add(ref err, DPath.Root, new DName("N"), DType.Number)
+                .Add(ref err, DPath.Root, new DName("E"), DType.Error);
+            type7 = type7.Add(ref err, DPath.Root.Append(new DName("Y")), new DName("F"), DType.Boolean)
+                .Add(ref err, DPath.Root.Append(new DName("Y")), new DName("G"), DType.Error);
+
+            Assert.Equal("*[field1:e, field2:n, field3:b, field4:*[field1:s, field2:s, field3:b, field4:e]]", type7.SafeDTypeToString());
+
+            type7 = type7.DropMulti(ref err, DPath.Root, new DName("N"), new DName("E"));
+            type7 = type7
+                .Drop(ref err, DPath.Root.Append(new DName("Y")), new DName("F"))
+                .Drop(ref err, DPath.Root.Append(new DName("Y")), new DName("G"));
+
+            type7 = type7.Drop(ref err, DPath.Root.Append(new DName("Q")), new DName("F"));
+
+            DType type12 = type1.Add(ref err, DPath.Root, new DName("Y"), DType.CreateTable(new TypedName(DType.Boolean, new DName("F")), new TypedName(DType.Error, new DName("E"))));
+            Assert.Equal("*[field1:n, field2:n, field3:n, field4:*[field1:e, field2:b]]", type12.SafeDTypeToString());
+
+            DType type8 = type1.Add(ref err, DPath.Root, new DName("X"), type6);
+            Assert.Equal("*[field1:n, field2:n, field3:n, field4:*[field1:b, field2:*[field1:s, field2:s]]]", type8.SafeDTypeToString());
+        }
+
+        [Fact]
+        public void EnumDTypeTests()
+        {
+            Assert.True(DType.TryParse("%n[A:0, B:1, C:2, D:3]", out DType type) && type.IsEnum);
+            Assert.True(DType.TryParse("%n[A:0, B:1, C:2, D:3]", out DType type2) && type2.IsEnum);
+        }
```

Comment:
Is it needed? There's no testing for the new function here.

Reply:
missing code :)